Add TFSpeech2Text

gante opened the pull request:
# What does this PR do?

This PR adds a TF port of Speech2Text. A summary of the changes:
- This model borrows a lot of code from TFBart, just like Speech2Text borrowed from Bart;
- Tried to follow the changes in other PRs to enable smooth interoperation with other parts of transformers (e.g. auto classes), might be missing a few things 👼 ;
- This seems to be the first TF model with speech as input, so I had to touch common TF code to enable correct data piping and misc operations (e.g. enable loading Conv1D PT weights into TF);
- Likewise, there were a few tests in `test_modelling_tf_common.py` that didn't quite fit this new kind of model.

TODO: 
- [x] create TF version of the weights, so we can load a model without `from_pt=True`

gante committed:
set the goal

gante committed:
Add wrapper classes

gante committed:
(tmp commit)

gante committed:
convert inner layers to tf

gante committed:
Add TF Encoder and Decoder layers

gante committed:
TFSpeech2Text models

gante committed:
Loadable model

gante committed:
TF model with same outputs as PT model

gante committed:
test skeleton

gante committed:
correct tests and run the fixup

gante committed:
revert some lines that shouldn't have been touched

gante commented:
Tagging @sgugger as a core dev, @patil-suraj as a core dev + original creator of our `Speech2Text`, and @Rocketknight1 as the TensorFlow boy. Feel free to redirect the reviews if you know of better people to review.

Some pipeline tests are failing almost surely due to the changes in `generation_tf_utils.py`, as some models expect `(encoder_outputs, past)` in the `past` variable and others don't -- having a look, but open to suggestions.

sgugger commented on src/transformers/generation_tf_utils.py:
```diff
@@ -870,6 +875,10 @@ def _generate_no_beam_search(
             )
 
         while cur_len < max_length:
+            # past drops the `encoder_outputs` during the loop and it may be needed (encoder-decoder models)
+            if len(past) > 1 and encoder_outputs is not None:
+                past = (encoder_outputs, past)
```
Let's double check with @patrickvonplaten for this change.

gante replied:
I've reworked this part to be less intrusive: the `past` variable is not touched, `encoder_outputs` is passed to `prepare_inputs_for_generation` a few lines below. This means that it will be ignored as part of `kwargs` in most models, and encoder-decoder models can now explicitly access it if they wish to do so (like `TFSpeech2Text`).

This change also solves the `assert` comments below, in `prepare_inputs_for_generation`.

sgugger commented on src/transformers/modeling_tf_pytorch_utils.py:
```diff
@@ -204,6 +206,11 @@ def load_pytorch_weights_in_tf2_model(tf_model, pt_state_dict, tf_inputs=None, a
             #    PT: (num_out_channel, num_in_channel, kernel[0], kernel[1])
             # -> TF: (kernel[0], kernel[1], num_in_channel, num_out_channel)
             array = numpy.transpose(array, axes=(2, 3, 1, 0))
+        elif transpose is TransposeType.CONV1D:
+            # Conv1D weight:
+            #    PT: (num_out_channel, num_in_channel, kernel)
+            # -> TF: (kernel, num_in_channel, num_out_channel)
+            array = numpy.transpose(array, axes=(2, 1, 0))
```
Very clean!

sgugger commented on src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:
```diff
@@ -0,0 +1,1591 @@
+        self.num_heads = num_heads
+        self.dropout = tf.keras.layers.Dropout(dropout)
+        self.head_dim = embed_dim // num_heads
+        assert self.head_dim * num_heads == self.embed_dim, "embed_dim must be divisible by num_heads"
```
We'll need to fix this assert in the original model (no new assert in the modeling files normally, but this is a copied from).

gante replied:
Copied the PT check (that raises a `ValueError`) to both models

gante replied:
Well, it turns out that many models used a copy of this 😅  Changed them all 👍

sgugger commented on src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:
```diff
@@ -0,0 +1,1591 @@
+        """
+        Args:
+            hidden_states (`tf.Tensor`): input to the layer of shape *(seq_len, batch, embed_dim)*
+            attention_mask (`tf.Tensor`): attention mask of size
+                *(batch, 1, tgt_len, src_len)* where padding elements are indicated by very large negative values.
+            encoder_hidden_states (`tf.Tensor`):
+                cross attention input to the layer of shape *(seq_len, batch, embed_dim)*
+            encoder_attention_mask (`tf.Tensor`): encoder attention mask of size
+                *(batch, 1, tgt_len, src_len)* where padding elements are indicated by very large negative values.
+            layer_head_mask (`tf.Tensor`): mask for attention heads in a given layer of size
+                *(decoder_attention_heads,)*
+            cross_attn_layer_head_mask (`tf.Tensor`): mask for heads of the cross-attention module.
+                *(decoder_attention_heads,)*
```
All the shapes should be in code blocks, not italics.

gante replied:
Copy-pasta from TFBart, changed in both 👍

sgugger commented on src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:
```diff
@@ -0,0 +1,1591 @@
+        hidden_states = residual + hidden_states
+
+        return (
+            hidden_states,
+            self_attn_weights,
+            cross_attn_weights,
+            present_key_value,
```
Night, this may fit in one line.

sgugger commented on src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:
```diff
@@ -0,0 +1,1591 @@
+                encoder_hidden_states=inputs["encoder_hidden_states"],
+                encoder_attention_mask=inputs["encoder_attention_mask"],
+                layer_head_mask=inputs["head_mask"][idx] if inputs["head_mask"] is not None else None,
+                cross_attn_layer_head_mask=inputs["cross_attn_head_mask"][idx]
+                if inputs["cross_attn_head_mask"] is not None
+                else None,
```
This is hard to read with the formatting, so put this in a separate line.

sgugger commented on src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:
```diff
@@ -0,0 +1,1591 @@
+        encoder_outputs=None,
+        **kwargs
+    ):
+        assert past is not None and len(past) in {1, 2}, f"past has to be an iterable of length 1,2 got {past}"
```
Please use a test and raise the appropriate Error :-)

sgugger commented on src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:
```diff
@@ -0,0 +1,1591 @@
+        **kwargs
+    ):
+        assert past is not None and len(past) in {1, 2}, f"past has to be an iterable of length 1,2 got {past}"
+        if len(past) == 1:
+            assert isinstance(past[0], tf.Tensor), f"`past[0]` has to be of type `tf.Tensor`, but is {type(past[0])}"
+            encoder_outputs = TFBaseModelOutput(last_hidden_state=past[0])
+            past_key_values = None
+        else:
+            assert (
+                len(past) == 2
+            ), "`past` has to be of length 2 with the encoder_outputs at the first position and past_key_values at the second position."
+            encoder_outputs, past_key_values = past
+            if isinstance(encoder_outputs, tuple):
+                assert isinstance(
+                    encoder_outputs[0], tf.Tensor
+                ), f"`encoder_outputs[0]` has to be of type `tf.Tensor`, but is {type(encoder_outputs[0])}"
+                encoder_outputs = TFBaseModelOutput(last_hidden_state=encoder_outputs[0])
+            elif isinstance(encoder_outputs, tf.Tensor):
+                encoder_outputs = TFBaseModelOutput(last_hidden_state=encoder_outputs)
+            assert (
+                past_key_values
+            ), f"decoder cached states must be truthy. got {past_key_values} from the 2nd element of past"
+            decoder_input_ids = decoder_input_ids[:, -1:]
+
+        assert isinstance(
+            encoder_outputs, TFBaseModelOutput
+        ), f"encoder_outputs should be a TFBaseModelOutput, Instead got {type(encoder_outputs)}."
```
Same here.

sgugger commented on tests/test_modeling_tf_common.py:
```diff
@@ -450,7 +452,16 @@ def test_compile_tf_model(self):
 
         for model_class in self.all_model_classes:
-            if self.is_encoder_decoder:
+            if model_class.__name__ in ["TFSpeech2TextModel", "TFSpeech2TextForConditionalGeneration"]:
+                inputs = {
+                    "decoder_input_ids": tf.keras.Input(
+                        batch_shape=(2, max_input),
+                        name="decoder_input_ids",
+                        dtype="int32",
+                    ),
+                    "input_ids": tf.keras.Input(batch_shape=(2, 32, max_input), name="input_ids", dtype="float32"),
+                }
```
Is there another (more general) check we could use here?

gante replied:
By more general do you mean instead of hardcoding the possible names in `model_class.__name__`? We have `TF_MODEL_FOR_SPEECH_SEQ_2_SEQ_MAPPING_NAMES`, but it doesn't cover `TFSpeech2TextModel` :(

sgugger commented on tests/test_modeling_tf_common.py:
```diff
@@ -995,10 +1011,10 @@ def test_lm_head_model_random_beam_search_generate(self):
-            else:
-                # num_return_sequences = 1
+            elif model_class.__name__ not in ["TFSpeech2TextForConditionalGeneration"]:
+                # Models with non-text inputs won't work here; num_return_sequences = 1
```
For those two tests, maybe we should overwrite the corresponding tests in the speech2text test file.

sgugger reviewed (approved):
Very nice addition, great work!

gante committed:
address PR comments

gante commented on src/transformers/generation_tf_utils.py:
```diff
@@ -756,6 +759,8 @@ def generate(
             )
             # expand encoder_outputs
             encoder_outputs = (tf.gather(encoder_outputs[0], expanded_batch_idxs, axis=0),)
+            if "attention_mask" in locals():  # vision models don't have this
```
Excellent point, for some reason I didn't notice that it was in the keyword arguments (and thus the current version would fail) 👍  Checking against `None` now

gante committed:
run make fixup

patrickvonplaten commented on src/transformers/generation_tf_utils.py:
```diff
@@ -756,6 +759,8 @@ def generate(
             # expand encoder_outputs
             encoder_outputs = (tf.gather(encoder_outputs[0], expanded_batch_idxs, axis=0),)
+            if attention_mask is not None:  # vision models don't have this
+                attention_mask = tf.gather(attention_mask, expanded_batch_idxs, axis=0)
```
Is this only needed for the new model or was it a bug to not have done this previously?

patrickvonplaten replied:
I'm a bit worried about this as it could lead to some unexpected errors in the more complicated models like TFRag. Could we maybe run the following tests here to double check that this change is ok:

`RUN_SLOW=1 pytest tests/test_modeling_tf_bart.py`
`RUN_SLOW=1 pytest tests/test_modeling_tf_t5.py`
`RUN_SLOW=1 pytest tests/test_modeling_tf_rag.py` (not that you need some more dependencies to be installed here)

gante replied:
Spot on, it failed for those models. 

I've reworked these lines to behave like pytorch's `_expand_inputs_for_generation()`, which is used in all modalities. I've tested the changes against bart, bert, t5, vit, and s2t -- let me know what you think, @patrickvonplaten

patrickvonplaten replied:
Cool! Could you please also run the slow `TFRag` tests - it's a bit annoying as it's a huge and slow test, but it's the model that is most prone to break if there are changes in tf generate

gante replied:
Can confirm it passes TFRag tests as well (had to spin up a larger machine 😅 )

patrickvonplaten replied:
Cool! Good for me then

patrickvonplaten commented on src/transformers/generation_tf_utils.py:
```diff
@@ -871,7 +877,12 @@ def _generate_no_beam_search(
+                past=past,
+                attention_mask=attention_mask,
+                use_cache=use_cache,
+                encoder_outputs=encoder_outputs,
```
If past and encoder_outputs is the same, why do we pass both here? Because the name `enocder_outputs` is cleaner? Would it maybe be better to keep `past` for now for consistency?

gante replied:
Once again, your comment led to insightful changes -- digging deeper, I found that the `past_key_values` output of the decoder is different on PT and TF (TF's is a tuple with the encoder's hidden states and past key values, and PT's is only the later). I had adapted the code from PT, so my `past_key_values` was missing the encoder outputs, and I had to pass them explicitly in order to work.

I've reverted this change and made TFSpeech2Text's `past_key_values` match the expected TF shape. An updated TF generate would probably have prevented this error 🤔

patrickvonplaten commented on tests/test_modeling_tf_speech_to_text.py:
```diff
@@ -0,0 +1,593 @@
+
+    def test_generation_librispeech_batched(self):
+        model = TFSpeech2TextForConditionalGeneration.from_pretrained(
+            "facebook/s2t-small-librispeech-asr", from_pt=True
```
Let's upload the models with TF weights no?

patrickvonplaten reviewed (commented):
Looks good to me in general! Maybe just two final things before merging:

- Quickly run some slow tests to make sure we haven't broken `generate` accidently for some of the more complicated models
- Should we maybe leave `past` for now until we improve the naming of generate more globally?

ToluClassics committed:
add t5 ner finetuning (#15432)

sgugger committed:
Add doc for add-new-model-like command (#15433)

NielsRogge committed:
Add header (#15434)

stas00 committed:
[deepspeed doc] fix import, extra notes (#15400)

* [deepspeed doc] fix import, extra notes

* typo

ydshieh committed:
Fix loss calculation in TFXXXForTokenClassification models (#15294)

* Fix loss calculation in TFFunnelForTokenClassification

* revert the change in TFFunnelForTokenClassification

* fix FunnelForTokenClassification loss

* fix other TokenClassification loss

* fix more

* fix more

* add num_labels to ElectraForTokenClassification

* revert the change to research projects

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

Rocketknight1 committed:
Fix spurious warning in TF TokenClassification models (#15435)

sgugger committed:
Change REALM checkpoint to new ones (#15439)

* Change REALM checkpoint to new ones

* Last checkpoint missing

patrickvonplaten committed:
[Trainer] suppress warning for length-related columns (#15421)

* [Trainer] suppress warning for length-related columns

* improve message

* Update src/transformers/trainer.py

patil-suraj committed:
[examples/Flax] add a section about GPUs (#15198)

* add a section about GPUs

* Apply suggestions from code review

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

ydshieh committed:
Fix TFLEDModel (#15356)

* fix tf led

* fix

* fix

* Add test_pt_tf_model_equivalence_extra for TFLED

* add a (temporary) test

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

patil-suraj committed:
correct positionla emb size (#15441)

patil-suraj committed:
[RobertaTokenizer] remove inheritance on GPT2Tokenizer (#15429)

* refactor roberta tokenizer

* refactor fast tokenizer

* remove old comment

Rocketknight1 committed:
Misfiring tf warnings (#15442)

* Fix spurious warning in TF TokenClassification models

* Fixing one last spurious warning

* Removing outdated warning altogether

itsTurner committed:
Add 'with torch.no_grad()' to BEiT integration test forward passes (#14961)

* Add 'with torch.no_grad()' to BEiT integration test forward pass

* Fix inconsistent use of tabs and spaces in indentation

peregilk committed:
Update modeling_wav2vec2.py (#15423)

* Update modeling_wav2vec2.py

With very tiny sound files (less than 0.1 seconds) the num_masked_span can be too long. The issue is described in issue #15366 and discussed with @patrickvonplaten.

* correct errors with mask time indices

* remove bogus file

* make fix-copies

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

sgugger committed:
Error when group_by_length is used with an IterableDataset (#15437)

patil-suraj committed:
skip test for XGLM (#15445)

stas00 committed:
[generate] fix synced_gpus default (#15446)

gante committed:
correct attention expansion

ydshieh committed:
remove "inputs" in tf common test script (no longer required) (#15262)

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

ydshieh committed:
Fix TF Causal LM models' returned logits (#15256)

* Fix TF Causal LM models' returned logits

* Fix expected shape in the tests

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

ydshieh committed:
fix from_vision_text_pretrained doc example (#15453)

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

patil-suraj committed:
[M2M100, XGLM] fix positional emb resize (#15444)

gante committed:
TFSpeech2Text pask_key_values with TF format

kamalkraj committed:
Update README.md (#15462)

fix typo

SaulLu committed:
replace assert with exception for padding_side arg in `PreTrainedTokenizerBase` `__init__` (#15454)

* replace assert with exception for `padding_side` arg in `PreTrainedTokenizerBase` `__init__`

* add test

* fix kwargs

* reformat test

* format

* format

* fix typo to render the documentation

SaulLu committed:
fix the `tokenizer_config.json` file for the slow tokenizer when a fast version is available (#15319)

* add new test

* update test

* remove `tokenizer_file` from `additional_files_names` in `tokenization_utils_base.py`

* add `tokenizer_file` for the fast only tokenizer

* change global variables layoutxml

* remove `"tokenizer_file"` from DPR tokenizer's Global variables

* remove `tokenizer_file` from herbert slow tokenizer init

* `"tokenizer_file"` from LED tokenizer's Global variables

* remove `tokenizer_file` from mbart slow tokenizer init

* remove `tokenizer_file` from slow tokenizer template

* adapt to versioning

* adapt the `test_tokenizer_mismatch_warning` test

* clean test

* clarify `VOCAB_FILES_NAMES` in tokenization_utils_fast.py

* Revert "remove `tokenizer_file` from mbart slow tokenizer init"

This reverts commit 0dbb723fa9c7599d4640fe30b3647a74eb4a64e1.

* Revert "`"tokenizer_file"` from LED tokenizer's Global variables"

This reverts commit 5a3f879bdd651233f3d74a3d1146c34cde82b0c2.

* Revert "remove `tokenizer_file` from herbert slow tokenizer init"

This reverts commit f5e10007b7b0ec5345e015b9de7ffec72c5407fd.

* Revert "remove `"tokenizer_file"` from DPR tokenizer's Global variables"

This reverts commit da0895330bedfafc81ae3073470a9348c669f032.

* set `tokenizer_file` in super `__init__` of mbart

gante committed:
delete commented line

gante committed:
revert previous messy commit

gante committed:
Merge remote-tracking branch 'origin/master' into add_tf_s2t

Rocketknight1 commented:
Cloned your branch and did some experimentation and LGTM! I noticed one issue - `model.save()` seems to encounter problems, but I'm not totally sure of why, or whether it's limited to this model or not. `save_pretrained` and `save_weights` worked correctly, and saving in `SavedModel` format has always been a bit shaky for us, so this isn't critical.

gante committed:
Fix pipeline tests; correct input variable name

patrickvonplaten commented on src/transformers/generation_tf_utils.py:
```diff
@@ -1131,8 +1119,9 @@ def _generate_beam_search(
         done = [False for _ in range(batch_size)]
 
         while cur_len < max_length:
+            # import pdb; pdb.set_trace()
```
can we delete this?

gante replied:
🤦

patrickvonplaten commented on src/transformers/generation_tf_utils.py:
```diff
@@ -715,23 +715,15 @@ def generate(
+            shape=(-1,),
+        )
+        # prepares text-based inputs
+        if len(shape_list(input_ids)) == 2:
```
Slightly worried here about the TF vision tests. Can you check this one as well:

```
RUN_SLOW=1 pytest tests/test_modeling_tf_vision_encoder_decoder.py
```

gante replied:
Can confirm that they pass 👍

patrickvonplaten reviewed (commented):
One final vision tests as mentioned before and then it's good for me as well

gante committed:
Fix dummy input sizes; Remove constant from nn.Parameters

patrickvonplaten commented on src/transformers/models/speech_to_text/modeling_speech_to_text.py:
```diff
@@ -153,9 +153,9 @@ def make_weights(self, num_embeddings: int, embedding_dim: int, padding_idx: Opt
             emb_weights = emb_weights.to(dtype=self.weights.dtype, device=self.weights.device)
 
-        self.weights = nn.Parameter(emb_weights)
+        self.weights = emb_weights
```
@patil-suraj - is this change ok? Why exactly do we have to do this here actually?

patrickvonplaten replied:
I'm not sure we can just save a tensor like this in PyTorch. When doing `model.to(device)` does this tensor correctly move to the device as well?

patrickvonplaten replied:
I don't really understand why this would cause an issue in TF<=>PT exactly - what was the issue? BTW, we always wrap the static sinusoidal encodings in a `nn.Parameter()` or `nn.Embedding(...)`, E.g. Vit's positional embeddings, DistilBERT's position embeddings. Why should we do it differently here?

gante replied:
> what was the issue?

This is the only model that I could find that may change the size of the embedding layer in the forward call ([here](https://github.com/huggingface/transformers/blob/master/src/transformers/models/speech_to_text/modeling_speech_to_text.py#L188)). The TF version for other embeddings allocates these named model parameters in [`build()`](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Layer#build), which is expected to happen before the first `call()`. Since `nn.Parameter()` is mostly a wrapper to add the variable to the model's list of parameters (AFAIK), then dropping it means we don't need a named parameter in TF, which means we can simplify the code and avoid solving the resizing issue.

sgugger replied:
Technically, someone who fine-tuned (or just saved and reshared) a Speecj2Text model will have this weight in the checkpoint. I don't know how it's used exactly, so I don't know if it's breaking to ignore it (we would have to put it in the `_keys_to_ignore_on_load_unexpected`).

gante replied:
(sorted on Slack -- conclusion: reverting this change since it is needed for the `to` method)

gante committed:
named embedding weights

gante commented:
@patrickvonplaten reverted the previous change and added the embedding weights as named variables in TF

gante committed:
run make fixup

patrickvonplaten reviewed (approved):
Thanks for working on it!
